staging/src/k8s.io/client-go/dynamic/simple.go

```diff
@@ -357,6 +357,7 @@ func (c *dynamicResourceClient) watchList(ctx context.Context, opts metav1.ListO
 		SpecificallyVersionedParams(&opts, dynamicParameterCodec, versionV1).
 		Timeout(timeout).
 		WatchList(ctx).
+		WithCustomDecoder(unstructured.UnstructuredJSONScheme).
```

Comment:
why is this needed?

Reply:
(Mentioned this out-of-band and adding it to the discussion here...)

I'm working on changing to the dynamic client to use the RESTClient's existing content-type negotiation functionality to select encoders and decoders -- removing its hardcoded assumptions about JSON -- and will have to address this. Shouldn't the negotiated embedded object decoder be used in all cases?

Reply:
> why is this needed?

The negotiated decoder is not able to decode into `UnstructuredList`. The `unstructured.UnstructuredJSONScheme` knows how to decode into `UnstructuredList`.

Reply:
Since no "into" argument is provided, the JSON decoder is falling back on the dynamic client's `ObjectCreater` which always returns `*Unstructured`. The UnstructuredJSONScheme uses a heuristic (presence of an "items" key) when no "into" argument is provided: https://github.com/kubernetes/kubernetes/blob/bded41a817a84cd7e51fa535ad6def319a892a3c/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/helpers.go#L392-L409.

The dynamic client's negotiated JSON decoder _will_ decode into an `*UnstructuredList`, but the current implementation here doesn't pass the destination `runtime.Object` to Decode (as you mention in a comment there). I ran a quick test of this on your branch locally and it does work.

Reply:
https://github.com/kubernetes/kubernetes/commit/e25714b9c9ff926fec56d9e66ef39959670c73e1